pkg/storage/unified/resource/metadata.go

```diff
@@ -0,0 +1,273 @@
+		return func(yield func(MetaDataObj, error) bool) {
+			yield(MetaDataObj{}, fmt.Errorf("resource is required"))
+		}
+	}
+	prefix, err := d.getPrefix(key)
+	if err != nil {
+		return func(yield func(MetaDataObj, error) bool) {
+			yield(MetaDataObj{}, err)
+		}
+	}
+	// List all keys in the prefix.
+	iter := d.kv.Keys(ctx, metaSection, ListOptions{
+		StartKey: prefix,
+		EndKey:   PrefixRangeEnd(prefix),
+		Sort:     SortOrderAsc,
+	})
+
+	return func(yield func(MetaDataObj, error) bool) {
+		var candidateKey *MetaDataKey // The current candidate key we are iterating over
+
+		// yieldCandidate is a helper function to yield results.
+		// Won't yield if the resource was last deleted.
+		yieldCandidate := func(key MetaDataKey) bool {
+			if key.Action == DataActionDeleted {
+				// Skip because the resource was last deleted.
+				return true
+			}
+			obj, err := d.kv.Get(ctx, metaSection, d.getKey(key))
+			if err != nil {
+				yield(MetaDataObj{}, err)
+				return false
+			}
+			var meta MetaData
+			value, err := io.ReadAll(obj.Value)
+			if err != nil {
+				yield(MetaDataObj{}, err)
+				return false
+			}
+			if err := json.Unmarshal(value, &meta); err != nil {
+				yield(MetaDataObj{}, err)
+				return false
+			}
+			return yield(MetaDataObj{
+				Key:   key,
+				Value: meta,
+			}, nil)
+		}
+
+		for key, err := range iter {
+			if err != nil {
+				yield(MetaDataObj{}, err)
+				return
+			}
+
+			metaKey, err := d.parseKey(key)
+			if err != nil {
+				yield(MetaDataObj{}, err)
+				return
+			}
+
+			if candidateKey == nil {
+				// Skip until we have our first candidate
+				if metaKey.ResourceVersion <= rv {
+					// New candidate found.
+					candidateKey = &metaKey
+				}
+				continue
+
+			}
+			// Should yield if either:
+			// - We reached the next key.
+			// - We reached a resource version greater than the target RV.
+			if !keyMatches(metaKey, *candidateKey) || metaKey.ResourceVersion > rv {
+				if !yieldCandidate(*candidateKey) {
+					return
+				}
+				// If we moved to a different resource and the current key is valid, make it the new candidate
+				if !keyMatches(metaKey, *candidateKey) && metaKey.ResourceVersion <= rv {
+					candidateKey = &metaKey
+				} else {
+					candidateKey = nil
+				}
+			} else {
+				// Update candidate to the current key (same resource, valid version)
+				candidateKey = &metaKey
+			}
+		}
+		if candidateKey != nil {
+			// Yield the last selected object
+			if !yieldCandidate(*candidateKey) {
+				return
+			}
+		}
+	}
+}
+
+func (d *metadataStore) Save(ctx context.Context, obj MetaDataObj) error {
+	valueBytes, err := json.Marshal(obj.Value)
+	if err != nil {
+		return err
+	}
+	return d.kv.Save(ctx, metaSection, d.getKey(obj.Key), io.NopCloser(bytes.NewReader(valueBytes)))
+}
+
+func (d *metadataStore) getKey(key MetaDataKey) string {
+	return fmt.Sprintf("%s/%s/%s/%s/%d~%s~%s", key.Group, key.Resource, key.Namespace, key.Name, key.ResourceVersion, key.Action, key.Folder)
```

Comment:
Do we need to escape group, resource, namespace or name to avoid extra `/` characters?